### ISSUE TITLE
TypeDeducer should fall back to docblock annotation when it returns mixed.

### ISSUE THREAD

**Issue by brad-jones** [php-integrator/core#56]
Again lets start with an example:

```php
class Foo
{
    /**
     * @var Bar
     */
    public $bar;
}

$foo = new Foo;

$foo->bar-> // I get completions here :)

$foo->bar = $something->that->is->a->bar->but->can->not->be->deduced();

$foo->bar-> // I don't get completions here :(
```

The deducer has returned "mixed", I think the deducer should fallback to whatever the docblock said in such cases.

**Comment by Gert-dev** [comment]
I think I understand your use case: you would like things such as autocompletion to still work with `Bar` as type because the last expression's type couldn't be deduced.

This poses a rather large issue for linting, more specifically the unknown member linting that is disabled by default in the linting package: currently, if I do this:

```php
$foo->bar = $something->that->is->a->bar->but->can->not->be->deduced();
$foo->bar->test(); // Error: expression "$foo->bar" does not have a type, so "test" can't be found.
```

If I fall back to the docblock and `Bar` happens to have a member `test`:

```php
$foo->bar = $something->that->is->a->bar->but->can->not->be->deduced();
$foo->bar->test(); // All is good, not making the programmer aware of any logic issue.
```

It would also be unexpected and feel confusing, as the type of "$foo->bar" at runtime must be whatever is returned by the expression of which the type could not be deduced. Anything else would pose a disconnect with what PHP itself does.

If, as programmer, you know what the expression will return, but static analysis is not smart enough to figure it out by itself, you can always resort to type override annotations [1].

[1] https://github.com/php-integrator/core/wiki/Hinting-And-Overriding-Types#type-override-annotations

**Comment by brad-jones** [comment]
Hmmm... I see where you are coming from.

Just turned on _"unknown member linting"_ and had a bit of play, I can see it's potential but right now as it says it's experimental and not very productive. Will be awesome once it's performance issues are sorted though. 

Perhaps only if the _"unknown member linting"_ is disabled then the docblock annotation is used as a fallback.

**Comment by Gert-dev** [comment]
That's true, it is indeed experimental because its performance is rather poor, especially on larger files. It is also not entirely correct yet (see also #36, #37, #38, #39 and others), which defeats the purpose of using it a bit. It seems incredibly useful for detecting subtle bugs when all those corner cases are ironed out.

As I mentioned earlier, falling back to the docblock annotation doesn't seem like a good idea in this case as it is not how PHP itself behaves. Not only would it require introducing additional complexity to the type deducer and propagating support for this everywhere in the codebase, but there will be other cases where you don't expect it to work like this, such as refactoring the name of a method. If the type of the variable is deduced in another way, renaming a method may incorrectly refactor the wrong method calls. 

I could then still apply this setting only for e.g. autocompletion, but this can confuse users as they could ask the question "The type is being deduced as being Bar here, then why isn't it doing X on Bar?".

In short, it feels a bit wrong to me to show autocompletion for `Bar`, as at that point `$foo->bar` has the type and value of whatever the type and value of the undeducable expression are, respectively. At runtime, to say it is a `Bar` would be incorrect. Static analysis should mimic runtime behavior as closely as possible (think of it as a substitute for a compile-time and link-time checking in compiled languages). You would also have no choice but to change the undeducable expression anyway since the call is not valid and will fail at runtime (so having no autocompletion is a hint that something is likely wrong).

**Comment by brad-jones** [comment]
Making an assumption your familiar with the Doctrine ORM.
Here is a more concrete example of the issue:

```php
class SomeDoctrineEntity
{
    /**
     * @ManyToOne(targetEntity="AnotherDoctrineEntity", inversedBy="someDoctrineEntities")
     * @var null|AnotherDoctrineEntity
     */
    public $anotherDoctrineEntity;
}

$aDoctrineEntity = new SomeDoctrineEntity;
$aDoctrineEntity->anotherDoctrineEntity = $anotherDoctrineCollection->matching(...SOME CRITERIA...)->first();
$aDoctrineEntity->anotherDoctrineEntity-> // it would be nice if this would auto complete
```

Ultimately PHP needs to get generics to solve this issue properly. In the end I actually wrote a script to generate proxies of ArrayCollection for each Entity. ie:  SomeDoctrineCollection, AnotherDoctrineCollection, FooCollection, BarCollection, etc Effectively faking generics in PHP.

All my entity properties are now type hinted against those Proxy Collections which allows PHP-Integrator to deduce the correct type. At runtime the proxy collection classes are simply ignored. It actully works amazingly well.

**Comment by Gert-dev** [comment]
I am familiar with Doctrine and see your point. As you mentioned, the correct solution to this problem would be generics. If that doesn't make it into PHP itself for some reason, I'm hoping PSR-5 will introduce some way to do it via docblocks [1]. (They are currently a bit uncertain about using the `<type>` syntax as it may be used by PHP itself if generics makes it in.)

I'm also eagerly awaiting (hoping for) generics in PHP myself, but as PHP integrator has to match PHP's behavior and the existing docblock standards (if you can even call them that) as closely as possible, I'd rather refrain from adding non-standard features. If they make it into PHP or a docblock standard, you can be certain I'll be implementing it, because I want it as well ;-).

[1] https://github.com/phpDocumentor/fig-standards/issues/89
